### PR TITLE
Fix vm containers could not restore after CRI-O restart

### PR DESCRIPTION
**Pull request by gozssky**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

When cri-o is running with kata runtime, cri-o could not restore containers correctly after it restarts. The `containerd-shim-kata-v2` and `qemu-system-x86_64` processes are leak every time.

This PR attempts to restore containers properly with the existing `containerd-shim-kata-v2` and `qemu-system-x86_64` processes. `containerd-shim-kata-v2` writes the gRPC socket address under container bundle path with a file named `address`. After cri-o restarts, we can read from that file for shim socket address and connect to. Besides, we also need to restore the container io so that we can exec or attach the container.

I tested these changes on our internal cluster, everything worked well.

#### Which issue(s) this PR fixes:

Fixes https://github.com/cri-o/cri-o/issues/2112 https://github.com/cri-o/cri-o/issues/5569

#### Does this PR introduce a user-facing change?

```release-note
Fix vm containers couldn't restore after cri-o restart
```

### CODE REVIEW

File: internal/oci/runtime_vm.go [hunk 1]
```diff
@@ -89,43 +91,11 @@ func newRuntimeVM(path, root, configPath string) RuntimeImpl {
...
-
-		opts = marshaledOtps
-	}
-
-	// First thing, we need to start the runtime daemon
-	if err := r.startRuntimeDaemon(ctx, c); err != nil {
-		return err
-	}
-
+func (r *runtimeVM) createContainerIO(ctx context.Context, c *Container, cioOpts ...cio.ContainerIOOpts) (_ *cio.ContainerIO, retErr error) {
```
Comment (haircommander) [inline internal/oci/runtime_vm.go:94]: nit: can we have this below the function that calls it?

File: internal/oci/runtime_vm.go [hunk 2]
```diff
@@ -164,9 +134,49 @@ func (r *runtimeVM) CreateContainer(ctx context.Context, c *Container, cgroupPar
...
+// CreateContainer creates a container.
+func (r *runtimeVM) CreateContainer(ctx context.Context, c *Container, cgroupParent string) (retErr error) {
+	log.Debugf(ctx, "RuntimeVM.CreateContainer() start")
+	defer log.Debugf(ctx, "RuntimeVM.CreateContainer() end")
+
+	// Lock the container
+	c.opLock.Lock()
+	defer c.opLock.Unlock()
+
+	// Lets ensure we're able to properly get construct the Options
```
Comment (haircommander) [inline internal/oci/runtime_vm.go:149]: this reads to me more of an operation to do when creating the runtime. WDYt about moving it there?

File: internal/oci/runtime_vm.go [hunk 3]
```diff
@@ -657,17 +667,75 @@ func (r *runtimeVM) UpdateContainerStatus(ctx context.Context, c *Container) err
 	return r.updateContainerStatus(ctx, c)
 }
 
+func (r *runtimeVM) restoreContainerIO(ctx context.Context, c *Container, state *task.StateResponse) error {
```
Comment (haircommander) [inline internal/oci/runtime_vm.go:670]: same here about defining below being called

File: internal/oci/runtime_vm.go [hunk 4]
```diff
@@ -680,6 +748,10 @@ func (r *runtimeVM) updateContainerStatus(ctx context.Context, c *Container) err
 		return errdefs.ErrNotFound
 	}
 
+	if err = r.restoreContainerIO(ctx, c, response); err != nil {
```
Comment (haircommander) [inline internal/oci/runtime_vm.go:751]: in the case where we haven't restarted, does this get called? should it?

**Comment by haircommander** [comment]
/ok-to-test
/area vm

Awesome @gozssky, many thanks for taking this on! I have some nits and questions, but in general I like the approach.

@fidencio @fgiudici PTAL

File: internal/oci/runtime_vm.go (hunk 3 above)
Comment (gozssky) [reply internal/oci/runtime_vm.go:670]: Do you mean put this function after `updateContainerStatus`?

File: internal/oci/runtime_vm.go (hunk 4 above)
Comment (gozssky) [reply internal/oci/runtime_vm.go:751]: yes, if container io has already existed in ctrs map, `restoreContainerIO` won't do more operations.

File: internal/oci/runtime_vm.go (hunk 2 above)
Comment (gozssky) [reply internal/oci/runtime_vm.go:149]: It seems ok, but this PR does n't change these lines.

Comment (haircommander) [reply internal/oci/runtime_vm.go:149]: ah nevermind then

File: internal/oci/runtime_vm.go (hunk 3 above)
Comment (haircommander) [reply internal/oci/runtime_vm.go:670]: yeah! and put `createContainerIO` below `restoreContainerIO` as well (as it calls `createContainerIO`)

Comment (gozssky) [reply internal/oci/runtime_vm.go]: done

**Comment by haircommander** [comment]
/retitle Fix vm containers could not restore after CRI-O restart

(ci doesn't handle apostrophes well...)

**Comment by fidencio** [comment]
Note to self: Take a look at this Tomorrow

File: internal/oci/runtime_vm.go (hunk 2 above)
Comment (fidencio) [reply internal/oci/runtime_vm.go]: This PR shuffles this code around, but this code was actually introduced to ensure we can configure the runtime to receive a specific configuration file.

**Review by fidencio — COMMENTED** [review]
@gozssky, first and foremost, this PR works like a charm, thanks for your contribution!

Now, a few tips just to improve the life of the reviewer for next interactions:
1. In https://github.com/cri-o/cri-o/pull/5574/commits/61f9d86d495343fa4356e660ed5731f78ffe9f4b you shuffle the `CreateContainer` around.  I´d leave that out of that commit, as that drives our eyes for a change there, while that's not the important part of your PR.
2. Move that code around as part of the second patch, please, as you're already moving code around there.
3. You're pulling in ctrio as a new dependency, which I don't see as a problem.  Historically, a long time ago, a bunch of containerd code was brought into CRI-O to avoid vendoring, which has proven to not be a good idea, so using it as a new dep as you did is **the correct way**.  Still on this, could you also check whether we still need something from `cio` / `cioutils`?  maybe everything we need already comes from `ctrio` and we could have **a follow-up patch** ensuring we use everything from there.

I'd appreciate if items 1 and 2 could be worked before getting this merged. 3 can come later.

Yet again, **nice work** @gozssky!

**Comment by gozssky** [comment]
> @gozssky, first and foremost, this PR works like a charm, thanks for your contribution!
> 
> Now, a few tips just to improve the life of the reviewer for next interactions:
> 
> 1. In [61f9d86](https://github.com/cri-o/cri-o/commit/61f9d86d495343fa4356e660ed5731f78ffe9f4b) you shuffle the `CreateContainer` around.  I´d leave that out of that commit, as that drives our eyes for a change there, while that's not the important part of your PR.
> 2. Move that code around as part of the second patch, please, as you're already moving code around there.
> 3. You're pulling in ctrio as a new dependency, which I don't see as a problem.  Historically, a long time ago, a bunch of containerd code was brought into CRI-O to avoid vendoring, which has proven to not be a good idea, so using it as a new dep as you did is **the correct way**.  Still on this, could you also check whether we still need something from `cio` / `cioutils`?  maybe everything we need already comes from `ctrio` and we could have **a follow-up patch** ensuring we use everything from there.
> 
> I'd appreciate if items 1 and 2 could be worked before getting this merged. 3 can come later.
> 
> Yet again, **nice work** @gozssky!

@fidencio Thanks for your review very much! I'm very sorry that I didn't check the code diff carefully before submitting this PR. But I'm not clear yet what I need to do for item 2. Could you explain a little more?

**Comment by fidencio** [comment]
> > @gozssky, first and foremost, this PR works like a charm, thanks for your contribution!
> > Now, a few tips just to improve the life of the reviewer for next interactions:
> > 
> > 1. In [61f9d86](https://github.com/cri-o/cri-o/commit/61f9d86d495343fa4356e660ed5731f78ffe9f4b) you shuffle the `CreateContainer` around.  I´d leave that out of that commit, as that drives our eyes for a change there, while that's not the important part of your PR.
> > 2. Move that code around as part of the second patch, please, as you're already moving code around there.
> > 3. You're pulling in ctrio as a new dependency, which I don't see as a problem.  Historically, a long time ago, a bunch of containerd code was brought into CRI-O to avoid vendoring, which has proven to not be a good idea, so using it as a new dep as you did is **the correct way**.  Still on this, could you also check whether we still need something from `cio` / `cioutils`?  maybe everything we need already comes from `ctrio` and we could have **a follow-up patch** ensuring we use everything from there.
> > 
> > I'd appreciate if items 1 and 2 could be worked before getting this merged. 3 can come later.
> > Yet again, **nice work** @gozssky!
> 
> @fidencio Thanks for your review very much! I'm very sorry that I didn't check the code diff carefully before submitting this PR. But I'm not clear yet what I need to do for item 2. Could you explain a little more?

@gozssky, ah, sorry, just add the `CreateContainer` code move as part of the patch that's already moving things around.
That's it and it should be good to go! :-)

**Comment by gozssky** [comment]
@fidencio I submitted the third patch https://github.com/cri-o/cri-o/pull/5574/commits/2e41e26dd5fc906f2d2ffef398e42441fba9b3d1.  I'm not sure if I understand your suggestion well. 😅

**Comment by fidencio** [comment]
@gozssky, my suggestion:
* On the commit https://github.com/cri-o/cri-o/pull/5574/commits/61f9d86d495343fa4356e660ed5731f78ffe9f4b, do **not** change the placement of the `CreateContainer` function.
* Instead, do that as part of https://github.com/cri-o/cri-o/pull/5574/commits/9c21a51e45a38d5f841686a8c4d52cf1134c2b15
* Drop https://github.com/cri-o/cri-o/pull/5574/commits/2e41e26dd5fc906f2d2ffef398e42441fba9b3d1, and let's have that as a different PR.

Sounds reasonable?

**Comment by gozssky** [comment]
> @gozssky, my suggestion:
> 
> * On the commit [61f9d86](https://github.com/cri-o/cri-o/commit/61f9d86d495343fa4356e660ed5731f78ffe9f4b), do **not** change the placement of the `CreateContainer` function.
> * Instead, do that as part of [9c21a51](https://github.com/cri-o/cri-o/commit/9c21a51e45a38d5f841686a8c4d52cf1134c2b15)
> * Drop [2e41e26](https://github.com/cri-o/cri-o/commit/2e41e26dd5fc906f2d2ffef398e42441fba9b3d1), and let's have that as a different PR.
> 
> Sounds reasonable?

@fidencio Thank you for the detailed explanation! I adjusted my two patches. The first patch 0731a9b5770bce105a147c218632f6e9102a6289 didn't change the `CreateContainer` function. And the second patched acde7255659e6a11a17f0bc35e4804b65cbb08a3 reuse the `createContainerIO` logic in `CreateContainer` and move functions after the function that calls them. Since the commit hash changed, I `force-push` my two commits. I'm not sure if this is appropriate.

**Comment by fidencio** [comment]
> @fidencio Thank you for the detailed explanation! I adjusted my two patches. The first patch [0731a9b](https://github.com/cri-o/cri-o/commit/0731a9b5770bce105a147c218632f6e9102a6289) didn't change the `CreateContainer` function. And the second patched [acde725](https://github.com/cri-o/cri-o/commit/acde7255659e6a11a17f0bc35e4804b65cbb08a3) reuse the `createContainerIO` logic in `CreateContainer` and move functions after the function that calls them. Since the commit hash changed, I `force-push` my two commits. I'm not sure if this is appropriate.

@gozssky, that's the way to go!
I will re-review / re-test it in the afternoon, thanks!

**Comment by fidencio** [comment]
@gozssky, integration tests are breaking, please take a look at those:
```
not ok 74 ctr update resources
# (from function `copyimg' in file ./helpers.bash, line 234,
#  from function `setup_img' in file ./helpers.bash, line 245,
#  from function `setup_crio' in file ./helpers.bash, line 255,
#  from function `start_crio' in file ./helpers.bash, line 337,
#  in test file ./ctr.bats, line 729)
#   `start_crio' failed
# time="2022-01-29T09:50:35Z" level=error msg="Error importing dir:/home/runner/work/cri-o/cri-o/.artifacts/redis-image: writing blob: adding layer with blob \"sha256:276d6b52cd5b0a1ec792aff4b4feb6969e60985969775a009b277104b9532396\": Error processing tar file(exit status 127): Inconsistency detected by ld.so: ../elf/dl-tls.c: 481: _dl_allocate_tls_init: Assertion `listp->slotinfo[cnt].gen <= GL(dl_tls_generation)' failed!\n"
# time="2022-01-29T09:50:37Z" level=fatal msg="connect: connect endpoint 'unix:///tmp/tmp.r3tXFgBgkr/crio.sock', make sure you are running as root and the endpoint has been started: context deadline exceeded"
# time="2022-01-29T09:50:39Z" level=fatal msg="connect: connect endpoint 'unix:///tmp/tmp.r3tXFgBgkr/crio.sock', make sure you are running as root and the endpoint has been started: context deadline exceeded"
```

**Review by fgiudici — COMMENTED** [review]
Nice patch, thanks @gozssky !
Cannot understand why tests fail: let me try to retest.

File: internal/oci/runtime_vm.go [hunk 5]
```diff
@@ -713,6 +695,96 @@ func (r *runtimeVM) updateContainerStatus(ctx context.Context, c *Container) err
...
+				return err
+			}
+		}
+		// Also try to remove the parent dir if it is empty.
+		for _, f := range iofiles {
+			_ = os.Remove(filepath.Dir(f))
+		}
+		return nil
+	}
+	_, err := r.createContainerIO(ctx, c, cio.WithFIFOs(ctrio.NewFIFOSet(cioCfg, closer)))
```
[the remark this answers is not in the document]
Comment (fidencio) [reply internal/oci/runtime_vm.go:738]: @fgiudici, I partially agree.

For a long time we decided to not directly vendor containerd code and we ended up actually copy and pasting code from containerd to our codebase, and as any copy & pasted code, those are not receiving bug fixes and are there to rot.
I'd be happy with directly vendor the code and get rid of everything we copied into our repos.  That would be cleaner and would esnure we actually get bug fixes from those.

WDYT?

[the remark this answers is not in the document]
Comment (gozssky) [reply internal/oci/runtime_vm.go:738]: > may be cleaner to keep all the code of fifo management in only one package 

@fgiudici I couldn't agree more. I'd like to do this in next PR.

**Comment by haircommander** [comment]
/approve

I tag @fidencio for the final lgtm

**Comment by fidencio** [comment]
I will run the last round of tests here and most likely have it merged sooner than later. :-)

**Comment by fidencio** [comment]
@gozssky, yet again, thanks a whole lot for the contribution!
/lgtm

**Pull request merged**